src/mitol/payment_gateway/api.py

```diff
@@ -306,3 +375,90 @@ def perform_processor_response_validation(self, request):
+        field names prepended with "req_". This will grab those fields
+        out of the response and convert it back to an Order and a set of
+        CartItems. (Not all payment processor APIs may support this, so this
+        is just defined for the CyberSource one.)
+
+        Args:
+            response: Dict; the data from the response
+
+        Returns:
+            Order
+        """
+        items = []
+
+        for i in range(0, int(response["req_line_item_count"])):
+            line = CartItem(
+                code=response[f"req_item_{i}_code"],
+                name=response[f"req_item_{i}_name"],
+                sku=response[f"req_item_{i}_sku"],
+                unitprice=response[f"req_item_{i}_unit_price"],
+                taxable=response[f"req_item_{i}_tax_amount"],
+                quantity=response[f"req_item_{i}_quantity"],
+            )
+            items.append(line)
+
+        order = Order(
+            username=response["req_consumer_id"],
+            ip_address=response["req_customer_ip_address"],
+            reference=response["req_reference_number"],
+            items=items,
+        )
+
+        return order
+
+    def decode_processor_response(self, request):
+        """
+        The CyberSource implementation of this. The ProcessorResponse
+        class was designed around this processor so the state codes map
+        one-to-one to the decision field that gets passed back from them.
+
+        The reason code and transaction ID fields don't appear in certain
+        states.
+        """
+        if request.method == "POST":
+            if hasattr(request, "data"):
+                response = request.data
+            else:
+                response = request.POST
+        else:
+            if hasattr(request, "query_params"):
+                response = request.query_params
+            else:
+                response = request.GET
+
+        fmt_response = ProcessorResponse(
+            ProcessorResponse.STATE_ERROR,
+            "Could not decode the processor response",
+            "",
+            "",
+        )
+
+        fmt_response.message = response["message"]
+
+        # CyberSource has no reason codes for cancellation at least
+        if "reason_code" in response:
+            fmt_response.response_code = response["reason_code"]
+
+        # Same for transaction_id
+        if "transaction_id" in response:
+            fmt_response.transaction_id = response["transaction_id"]
+
+        if response["decision"] == "ACCEPT":
+            fmt_response.state = ProcessorResponse.STATE_ACCEPTED
+        elif response["decision"] == "DECLINE":
+            fmt_response.state = ProcessorResponse.STATE_DECLINED
+        elif response["decision"] == "ERROR":
+            fmt_response.state = ProcessorResponse.STATE_ERROR
+            # maybe should log here? this is a straight-up something went wrong between the app and the processor state
```

Comment:
Is the application integrating with this likely to have more contextual information that would be useful for this logging message? If not, we should definitely log an error here.